di/main.py

```diff
@@ -112,11 +112,20 @@ def __init__(self, config=ConfigManager):
 
     def wrapper(fn, warn=True):
         # Extract default values for keyword arguments
-        args, varargs, keywords, defaults = inspect.getargspec(fn)
-        if defaults:
-            defaults = dict(zip(reversed(args), reversed(defaults)))
-        else:
-            defaults = {}
+        try:
+            args, _ , _ , defaults = inspect.getargspec(fn)[:4]
+            if defaults:
+                defaults = dict(zip(reversed(args), reversed(defaults)))
+            else:
+                defaults = {}
+        except ValueError:
+            if not PY2:
+                # try to got it from kwonlydefaults as a sort of functions
+                # with a signature such as fn(*args, di=di_value, **kwargs)
+                # defaults come well packed
+                defaults = inspect.getfullargspec(fn)[5]
+                if not defaults:
+                    defaults = {}
```

Comment:
Yea, the PR was only an attempt without a solid coverage of all cases. Your approximation looks solid and seems to cover all the cases working with a mixture of PY3 and PY2 giving priority to the new `signature` method.

This next week, if it is not already done by you, I can make a proper PR with the test coverage expected.

does it sound good for you ?